File: day7.1.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <string>
#include <stdexcept>

using namespace std;

typedef unsigned long long u64;

bool valid(u64 targ_sum, const vector<u64>& operands, u64 curr_sum, int i) {
    if (i == operands.size()) {
        return curr_sum == targ_sum;
    }
    if (curr_sum > targ_sum) return false;
    return valid(targ_sum, operands, curr_sum + operands[i], i + 1) ||
           valid(targ_sum, operands, curr_sum * operands[i], i + 1);
}

u64 process_line(const string& line) {
    istringstream iss(line);
    u64 sum;
    char colon;
    iss >> sum >> colon; 

    vector<u64> operands;
    string operand;
    while (iss >> operand) {
        if (!operand.empty()) {
            operands.push_back(stoll(operand));
        }
    }
    
    return valid(sum, operands, operands[0], 1) ? sum : 0;
}

int main() {
    ifstream file("input7.2.txt");
    if (!file.is_open()) {
        cerr << "Could not open file" << endl;
        return 1;
    }
    
    auto total = 0l;
    string line;
    while (getline(file, line)) {
        total += process_line(line);
    }
    cout << total << endl;
    return 0;
}

File: day7.1.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

class Solution {
    public static void main(String[] args) throws FileNotFoundException {
        File txt = new File("input7.2.txt");
        var total = 0l;
        try (Scanner scan = new Scanner(txt)) {
            while (scan.hasNextLine()) {
                var line = scan.nextLine();
                total += processLine(line);
            }
        }
        System.out.println(total);
    }

    private static long processLine(String line) {
        var parts = line.split(":");
        var sum = Long.parseLong(parts[0]);
        var operands = new ArrayList<Long>();
        for (var operand : parts[1].split(" ")) {
            if (operand.isBlank()) continue;
            operands.add(Long.parseLong(operand));
        }
        return isValid(sum, operands, operands.get(0), 1) ? sum : 0;
    }

    private static boolean isValid(long targetSum, List<Long> operands, long currentSum, int i) {
        if (i == operands.size()) {
            return currentSum == targetSum;
        }
        if (currentSum > targetSum) return false;
        return isValid(targetSum, operands, currentSum + operands.get(i), i + 1) ||
             isValid(targetSum, operands, currentSum * operands.get(i), i + 1);
    }
}
